packages/api/test/e2e/doughnut.e2e.ts

```diff
@@ -0,0 +1,206 @@
+// Copyright 2020 Centrality Investments Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {hexToU8a, assert} from '@polkadot/util';
+import {CENNZnut as encodeCennznut} from '@cennznet/cennznut-wasm';
+import {Doughnut as DoughnutMaker} from '@plugnet/doughnut-wasm';
+import {Api} from '../../src/Api';
+import {Keypair} from '@plugnet/util-crypto/types';
+import {Keyring} from '@polkadot/api';
+import {KeyringPair} from '@plugnet/keyring/types';
+import initApiPromise from '../../../../jest/initApiPromise';
+import {cryptoWaitReady} from '@plugnet/util-crypto';
+import Doughnut from '@cennznet/types/Doughnut';
+
+/// Helper for creating CENNZnuts which takes module/section name and method name to be same for doughnut permissioning
+function makeCennznut(module: string, method: string): Uint8Array {
+    const moduleVec = [
+        [
+            module, {
+            'name': module,
+            'methods': [
+                [
+                    method, {
+                    'name': method,
+                    'constraints': null,
+                },
+                ],
+            ],
+        },
+        ],
+    ];
+    const contractVec = [];
+    const cennznut = new encodeCennznut(moduleVec, contractVec);
+    return cennznut.encode();
+}
+
+/// Helper for creating v0 Doughnuts where user can pass expiry time for doughnut to be valid and 'not before' which would mean doughnut is not applied before this block
+function makeDoughnut(
+    issuer: Keypair,
+    holder: KeyringPair,
+    permissions: Record<string, Uint8Array>,
+) {
+    const d = new DoughnutMaker(
+        issuer.publicKey,
+        holder.publicKey,
+        (new Date().getTime()) * 2, // expiry set to twice the current timestamp
+        1, //notBefore
+    );
+    for (const key in permissions) {
+        d.addDomain(key, permissions[key]);
+    }
+    d.signSr25519(new Uint8Array(issuer.secretKey));
+    return d.encode();
+}
+
+
+const Dave = {
+    address: '5DAAnrj7VHTznn2AWBemMuyBwZWs6FNFjdyVXUeYum3PTXFy',
+    uri: '//Dave',
+};
+
+describe('Doughnut for CennznetExtrinsic', () => {
+    let aliceKeyPair = {
```

Comment:
Can we give a different name to this pair so not to be mistaken for alice?

Reply:
this key pair if for Alice.. This is Alice's secret key..so its better with aliceKeyPair

Reply:
If it's indeed alice, then we shouldn't get alice separately from a library. Remember if we update keyring library there is a chance that it returns alice with a different secret key.

Reply:
as discussed